design/authenticators/authn_gcp/authn_gcp_solution_design.md

```diff
@@ -0,0 +1,312 @@
+
+## Solution
+We will add a new GCP authenticator to Conjur, which GCE hosts will can authenticate with by delivering their host Conjur identity and google identity as JWT token.
+
+As mentioned in the [feature doc](https://ca-il-confluence.il.cyber-ark.com/display/rndp/Conjur+GCP+Authenticator#/), before any authentication 
+request is sent to Conjur, the admin will load the authenticator policy: 
+
+```yaml
+# policy id needs to match the convention `conjur/authn-azure/<service ID>`
+- !policy
+  id: conjur/authn-gcp/mygcp
+  body:
+  - !webservice
+      
+  - !group apps
+  
+  - !permit
+    role: !group apps
+    privilege: [ read, authenticate ]
+    resource: !webservice
+```
+
+All the following configurations will have constants values and will not require any action from the customer:
+
+* **provider-uri** = `https://www.googleapis.com/oauth2/v1/certs` #represnts the address to fetch the public key in order to validate the JWT token
+* **mandatory-claims** = `iss, exp, aud, iat, nbf` #the list of the mandatory claims to validate in the JWT token
+* **iss** = `https://accounts.google.com` #the issuer claim we should validate in the JWT token
+* **aud** = `conjur` #the audience claim we should validate in the JWT token
+
+### GCP Resource Restrictions
+To authenticate with Conjur, GCP-specific fields will need to be provided in the host annotations of the Conjur host identity with the following options:
+
+* **project_id** - A customizable unique identifier for your project.
+* **instance_name** - The name of the GCE machine
+* **service_account_id** - The google instance identity   
+* **service_account_email** - The email of the instance identity
+
+At least one of the above should be provided, if more than 1 is provided, we will authenticate them with AND logic.
+
+A Conjur host will be defined as follows with their annotations holding GCP-specific identification properties.
+
+```yaml
+- !policy
+  id: <policy-id>
+  body:
+    - !group
+ 
+    - &hosts
+      - !host
+        id: myapp
+        annotations:
+          authn-gcp/instance-name: <instance-name>
+          authn-gcp/project-id: <project-id>
+          authn-gcp/service-account-id: <service-account-id>
+          authn-gcp/service-account-email: <service-account-email>
+           
+    - !grant
+      role: !group
+      members: *hosts
+          
+- !grant
+  role: !group /conjur/authn-gcp/mygcp/apps
+  member: !group <policy-id>
+```
+
+### Environment-provided service account google identity
+In order to authenticate against Conjur GCP authenticator, a GCE instance first need to fetch its environment-provided identity.
+
+**Example:** 
+Run the following from the GCE machine:
+
+`curl \ --header Metadata-Flavor: Google" \ --get \ --data-urlencode "audience=conjur" \ --data-urlencode "format=full" \ "http://metadata/computeMetadata/v1/instance/service-accounts/default/identity`
+
+The audience value must to be `conjur` 
```

Comment:
please, explain why 'conjur'

Reply:
Again I suggest to add the host id in Conjur to the audience claim.

Reply:
@oburstein-hub - would appreciate your input here - do we want to have a standard audience structure across our services?

Reply:
yes - I prefer to have standard audience structure for all our services, and its values should be 'conjur'
@InbalZilberman  - please approve

Reply:
@InbalZilberman @Tovli @sashaCher  @oburstein-hub 
By specifying the `http://conjur/account/host/my-policy/my-host/authenticate` in the audience claim same as you do in Conjure's authenticate URL `http://conjur/account/host/my-policy/my-host/authenticate` you are binding the token to its destination url, that is also what Google [recommends](https://cloud.google.com/compute/docs/instances/verifying-instance-identity).
```
[AUDIENCE] is the unique URI agreed upon by both the instance and the system verifying the instance's identity. For example, the audience could be a URL for the connection between the two systems.
```

Reply:
@eran - please elaborate what is the advantage of having url-formatted audience. As per my understand - it means that audience  will be coupled with specific conjur operation URL, right